### ISSUE TITLE
Handle multiple "stack" nodes for errors in Valgrind XML report files

### ISSUE THREAD

**Issue by jhoussay** [SonarOpenCommunity/sonar-cxx#1440]
Hi,

For some Valgrind errors (such as "InvalidWrite"), the corresponding XML node contains two distinct "stack" nodes, one corresponding to the allocation stack and the other one corresponding to the memory misuse stack.

Unfortunately, sonar cxx only presents the content of one "stack'" node found in the "error" node. In my case, this is pretty annoying as the stack available in Sonar is the allocation stack and does not allow easy identication of the memory misuse.

It seems to come from the `parseErrorTag `method in [ValgrindReportParser.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindReportParser.java). When multiple "stack" nodes are contained in the "error" node, only the last one is kept:
```java
String kind = null;
String text = null;
ValgrindStack stack = null;
while (child.getNext() != null) {
   String tagName = child.getLocalName();
    if ("kind".equalsIgnoreCase(tagName)) {
      kind = child.getElemStringValue();
    } else if ("xwhat".equalsIgnoreCase(tagName)) {
      text = child.childElementCursor("text").advance().getElemStringValue();
    } else if ("what".equalsIgnoreCase(tagName)) {
      text = child.getElemStringValue();
    } else if ("stack".equalsIgnoreCase(tagName)) {
      stack = parseStackTag(child);
    }
}

[...]

return new ValgrindError(kind, text, stack);
```

Handling multliple stacks at this point would require to do so in [ValgrindError.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java) as well and modify the `getLastOwnFrame` method in order to provide the "best" stack origin...

Julien

**Comment by guwirth** [comment]
@jhoussay could you
* provide an XML report sample please. 
* Is this one message with two locations or two messages? Please have also a look to https://github.com/SonarOpenCommunity/sonar-cxx/pull/1436. Similar to this?

**Comment by jhoussay** [comment]
Hi @guwirth ,

Here is an example of report : [valgrind_output.xml.txt](https://github.com/SonarOpenCommunity/sonar-cxx/files/1846742/valgrind_output.xml.txt).

The invalid write "error" node starts line 37. It contains two "stack" nodes, the first one (line 42) corresponding to the invalid write itself (2) and the second one (line 141) corresponding to the memory allocation (1).
1. ```float *pToto = new float[10]; ```
2. ```pToto[10] = 10.f```

As I said in the first post, only the last stack is kept and, in SonarQube, I have the following behaviour, only pointing to the allocation place, which does not really help to spot the invalid memory access.
![image](https://user-images.githubusercontent.com/37700528/37892681-e74508c8-30d8-11e8-9683-1c88fae7b570.png)

So, to my mind, we need one message with the two locations provided in the "error" node.


I had a look to #1436, I would say it corresponds to the same issue but for cppcheck output files. As far as I understand sonar cxx, maybe this pull request could be used for valgrind report parsing as well : 

- The [ValgrindError.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java) class could be modified in order to store one or more locations,

- The `parseErrorTag` method in [ValgrindReportParser.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindReportParser.java) should handle multple "stack" nodes and provide them as multiple locations to the `ValgrindError` instance returned by the method,

- In [CxxValgrindSensor.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/CxxValgrindSensor.java), the new `saveUniqueLocation` provided by the PR could be used.

**Comment by ivangalkin** [comment]
@guwirth @jhoussay #1447 has to address stack frames of Valgrind finding: all frames, which belong to the analyzed project will be referenced as secondary locations.

An other issue I can recognize from the screenshot is the terrible formatting of the stack trace. I guess it should be enough to insert it as a code block. Markdown is supported in comments, not sure about issues. It must be tested.

```diff
diff --git a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/CxxValgrindSensor.java b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/CxxValgrindSensor.java
index 160dcfc..33d5916 100644
--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/CxxValgrindSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/CxxValgrindSensor.java
@@ -73,7 +73,7 @@ public class CxxValgrindSensor extends CxxReportSensor {
       ValgrindFrame frame = error.getLastOwnFrame(context.fileSystem().baseDir().getPath());
       if (frame != null) {
         saveUniqueViolation(context, CxxValgrindRuleRepository.KEY,
-          frame.getPath(), frame.getLine(), error.getKind(), error.toString());
+          frame.getPath(), frame.getLine(), error.getKind(), error.toMarkdown());
       } else {
         LOG.warn("Cannot find a project file to assign the valgrind error '{}' to", error);
       }
diff --git a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java
index e38acb9..b9f890f 100644
--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/valgrind/ValgrindError.java
@@ -49,6 +49,10 @@ class ValgrindError {
     return text + "\n\n" + stack;
   }
 
+  public String toMarkdown() {
+    return text + "\n\n```\n" + stack + "\n```\n";
+  }
+
   @Override
   public boolean equals(Object o) {
     if (this == o) {
```

**Comment by jhoussay** [comment]
Hi @guwirth, @ivangalkin,

I just relaunched the analysis with #1447.
I saw a little difference between the line number and the statement : a "1" is displayed, but I still was not able to find the real location of my invalid memory access.
![image](https://user-images.githubusercontent.com/37700528/38744148-4c46391a-3f41-11e8-8d51-4aec354246dd.png)

I have the feeling that multiple code locations are enabled (as seen at the bottom of the following screenshot) but are not handled in the valgrind report.
![image](https://user-images.githubusercontent.com/37700528/38744718-d42bf760-3f42-11e8-929d-a4358adbda62.png)

PS: @ivangalkin, I clearly agree that a better formatting for valgrind call stacks would be very nice :-)
